### ISSUE TITLE
firewallcmd-ipset with ipv6

### ISSUE THREAD

**Issue by adbario** [fail2ban/fail2ban#1990]
Fail2Ban version: 0.10.1 from GitHub
OS: CentOS 7.4 

When using banaction firewallcmd-ipset and ipv6, there's an error on log:
```
ipset v6.29: Syntax error: cannot parse xxxx:xxx:xxx:xxxx::xxx: resolving to IPv4 address failed
```

This seems to be because actionstart creates only ipset for ipv4 offenders. Is there any way around this or should firewallcmd-new be used with ipv6, even though I guess it uses more resources when ban list gets bigger?

**Comment by adbario** [comment]
Hi, and sorry about the missing information. Other than jail.conf, there's no changes from default 0.10.1 version.

```
# jail.local

[DEFAULT]
bantime   = 3600
findtime  = 600
maxretry  = 3
banaction = firewallcmd-ipset

[sshd]
enabled = true
mode    = aggressive
backend = systemd
```

Banning ipv4 connections work fine, but with ipv6 ther error log looks like this:
```
2017-12-05 21:21:12,227 fail2ban.utils          [1103]: Level 39 1e0d030 -- exec: ipset add f2b-sshd6 xxxx:xxxx:x:xxx::x timeout 3600 -exist
2017-12-05 21:21:12,228 fail2ban.utils          [1103]: ERROR   1e0d030 -- stderr: 'ipset v6.29: Syntax error: cannot parse xxxx:xxxx:x:xxx::x: resolving to IPv4 address failed'
2017-12-05 21:21:12,228 fail2ban.utils          [1103]: ERROR   1e0d030 -- returned 1
2017-12-05 21:21:12,228 fail2ban.actions        [1103]: ERROR   Failed to execute ban jail 'sshd' action 'firewallcmd-ipset' info 'ActionInfo({'ipfailures': 3, 'ip-rev': '5.0.0.0.0.0.0.0.0.0.0.0.0.0.0.0.4.9.7.0.b.0.0.0.0.c.1.1.0.0.a.2.', 'family': 'inet6', 'ipmatches': '', 'matches': '', 'ip': 'xxxx:xxxx:x:xxx::x', 'ipjailmatches': '', 'ipjailfailures': 3, 'F-*': {'matches': [], 'failures': 3, 'ip6': 'xxxx:xxxx:x:xxx::x', 'mlfid': 'domain.com sshd[1519]: ', 'user': 'xxxx'}, 'fid': 'xxxx:xxxx:x:xxx::x', 'time': 1512508871.905542, 'failures': 3, 'restored': 0, 'ip-host': None})': Error banning xxxx:xxxx:x:xxx::x
```

With firewallcmd-new it all works fine, but would rather use firewallcmd-ipset to save some resources.

```
$ fail2ban-client -d

['set', 'syslogsocket', 'auto']
['set', 'loglevel', 'INFO']
['set', 'logtarget', '/var/log/fail2ban.log']
['set', 'dbfile', '/var/lib/fail2ban/fail2ban.sqlite3']
['set', 'dbpurgeage', '1d']
['add', 'sshd', 'systemd']
['set', 'sshd', 'maxlines', 1]
['set', 'sshd', 'prefregex', '^<F-MLFID>(?:\\[\\])?\\s*(?:<[^.]+\\.[^.]+>\\s+)?(?:\\S+\\s+)?(?:kernel: \\[ *\\d+\\.\\d+\\]\\s+)?(?:@vserver_\\S+\\s+)?(?:(?:(?:\\[\\d+\\])?:\\s+[\\[\\(]?sshd(?:\\(\\S+\\))?[\\]\\)]?:?|[\\[\\(]?sshd(?:\\(\\S+\\))?[\\]\\)]?:?(?:\\[\\d+\\])?:?)\\s+)?(?:\\[ID \\d+ \\S+\\]\\s+)?</F-MLFID>(?:(?:error|fatal): (?:PAM: )?)?<F-CONTENT>.+</F-CONTENT>$']
['multi-set', 'sshd', 'addfailregex', ['^[aA]uthentication (?:failure|error|failed) for <F-USER>.*</F-USER> from <HOST>( via \\S+)?\\s*(?: \\[preauth\\])?\\s*$', '^User not known to the underlying authentication module for <F-USER>.*</F-USER> from <HOST>\\s*(?: \\[preauth\\])?\\s*$', '^Failed \\S+ for (?P<cond_inv>invalid user )?<F-USER>(?P<cond_user>\\S+)|(?(cond_inv)(?:(?! from ).)*?|[^:]+)</F-USER> from <HOST>(?: port \\d+)?(?: on \\S+(?: port \\d+)?)?(?: ssh\\d*)?(?(cond_user): |(?:(?:(?! from ).)*)$)', '^<F-USER>ROOT</F-USER> LOGIN REFUSED.* FROM <HOST>\\s*(?: \\[preauth\\])?\\s*$', '^[iI](?:llegal|nvalid) user <F-USER>.*?</F-USER> from <HOST>(?: port \\d+)?(?: on \\S+(?: port \\d+)?)?\\s*$', '^User <F-USER>.+</F-USER> from <HOST> not allowed because not listed in AllowUsers\\s*(?: \\[preauth\\])?\\s*$', '^User <F-USER>.+</F-USER> from <HOST> not allowed because listed in DenyUsers\\s*(?: \\[preauth\\])?\\s*$', '^User <F-USER>.+</F-USER> from <HOST> not allowed because not in any group\\s*(?: \\[preauth\\])?\\s*$', '^refused connect from \\S+ \\(<HOST>\\)\\s*(?: \\[preauth\\])?\\s*$', '^Received <F-MLFFORGET>disconnect</F-MLFFORGET> from <HOST>(?: port \\d+)?(?: on \\S+(?: port \\d+)?)?:\\s*3: .*: Auth fail(?: \\[preauth\\])?\\s*$', '^User <F-USER>.+</F-USER> from <HOST> not allowed because a group is listed in DenyGroups\\s*(?: \\[preauth\\])?\\s*$', "^User <F-USER>.+</F-USER> from <HOST> not allowed because none of user's groups are listed in AllowGroups\\s*(?: \\[preauth\\])?\\s*$", '^pam_unix\\(sshd:auth\\):\\s+authentication failure;\\s*logname=\\S*\\s*uid=\\d*\\s*euid=\\d*\\s*tty=\\S*\\s*ruser=<F-USER>\\S*</F-USER>\\s*rhost=<HOST>\\s.*(?: \\[preauth\\])?\\s*$', '^(error: )?maximum authentication attempts exceeded for <F-USER>.*</F-USER> from <HOST>(?: port \\d+)?(?: on \\S+(?: port \\d+)?)?(?: ssh\\d*)?(?: \\[preauth\\])?\\s*$', '^User <F-USER>.+</F-USER> not allowed because account is locked(?: \\[preauth\\])?\\s*', '^<F-MLFFORGET>Disconnecting</F-MLFFORGET>: Too many authentication failures(?: for <F-USER>.+?</F-USER>)?(?: \\[preauth\\])?\\s*', '^<F-NOFAIL>Received <F-MLFFORGET>disconnect</F-MLFFORGET></F-NOFAIL> from <HOST>: 11:', '^<F-NOFAIL>Connection <F-MLFFORGET>closed</F-MLFFORGET></F-NOFAIL> by <HOST>(?: \\[preauth\\])?\\s*$', '^Did not receive identification string from <HOST>(?: \\[preauth\\])?\\s*$', '^Connection <F-MLFFORGET>reset</F-MLFFORGET> by <HOST>(?: port \\d+)?(?: on \\S+(?: port \\d+)?)?(?: \\[preauth\\])?\\s*', '^<F-NOFAIL>SSH: Server;Ltype:</F-NOFAIL> (?:Authname|Version|Kex);Remote: <HOST>-\\d+;[A-Z]\\w+:', '^Read from socket failed: Connection <F-MLFFORGET>reset</F-MLFFORGET> by peer(?: \\[preauth\\])?\\s*', '^Received <F-MLFFORGET>disconnect</F-MLFFORGET> from <HOST>(?: port \\d+)?(?: on \\S+(?: port \\d+)?)?:\\s*14: No supported authentication methods available(?: \\[preauth\\])?\\s*$', '^Unable to negotiate with <HOST>(?: port \\d+)?(?: on \\S+(?: port \\d+)?)?: no matching (?:cipher|key exchange method) found.', '^Unable to negotiate a (?:cipher|key exchange method)(?: \\[preauth\\])?\\s*$', '^<F-NOFAIL>Connection from</F-NOFAIL> <HOST>']]
['set', 'sshd', 'datepattern', '{^LN-BEG}']
['set', 'sshd', 'addjournalmatch', '_SYSTEMD_UNIT=sshd.service', '+', '_COMM=sshd']
['set', 'sshd', 'usedns', 'warn']
['set', 'sshd', 'maxretry', 3]
['set', 'sshd', 'logencoding', 'auto']
['set', 'sshd', 'bantime', '3600']
['set', 'sshd', 'ignorecommand', '']
['set', 'sshd', 'findtime', '600']
['set', 'sshd', 'addaction', 'firewallcmd-ipset']
['multi-set', 'sshd', 'action', 'firewallcmd-ipset', [['actionunban', 'ipset del <ipmset> <ip> -exist'], ['actionstop', 'firewall-cmd --direct --remove-rule <family> filter INPUT 0 -p tcp -m multiport --dports ssh -m set --match-set <ipmset> src -j REJECT --reject-with <rejecttype>\nipset flush <ipmset>\nipset destroy <ipmset>'], ['actionstart', 'ipset create <ipmset> hash:ip timeout <bantime>\nfirewall-cmd --direct --add-rule <family> filter INPUT 0 -p tcp -m multiport --dports ssh -m set --match-set <ipmset> src -j REJECT --reject-with <rejecttype>'], ['actionban', 'ipset add <ipmset> <ip> timeout <bantime> -exist'], ['protocol', 'tcp'], ['family', 'ipv4'], ['ipmset', 'f2b-<name>'], ['port', 'ssh'], ['chain', 'INPUT'], ['service', 'ssh'], ['actname', 'firewallcmd-ipset'], ['zone', 'public'], ['bantime', '3600'], ['rich-blocktype', "reject type='<rejecttype>'"], ['rejecttype', 'icmp-port-unreachable'], ['rejecttype?family=inet6', 'icmp6-port-unreachable'], ['blocktype', 'REJECT --reject-with <rejecttype>'], ['ipmset?family=inet6', 'f2b-<name>6'], ['name', 'sshd'], ['family?family=inet6', 'ipv6']]]
```

**Comment by sebres** [comment]
Hmm... strange.
The adding of IP address (different families) to ipset don't differ in the command line (are the same, just another set-name used). 
Below you'll find the example in our test-cases:
https://github.com/fail2ban/fail2ban/blob/0.10/fail2ban/tests/servertestcase.py#L1675

As regard the ipset command line (that fails) it looks pretty good and set-name is from IPv6-family set
 (`f2b-sshd6`, corresponding IPv4 set called `f2b-sshd`, without 6).
Never seen up now... Looks like a bug in your ipset version...

Are you sure that `xxxx:xxxx:x:xxx::x` (blacked out?) is really an IPv6 address?
Possibly we've a filter issue (or something else).

**Comment by sebres** [comment]
Found a bug in the action definition: the create of set missing `family inet6` at end.

**Comment by adbario** [comment]
Yes that's and IPv6 address, same thing happens if I use the example from test-case:
```
$ ipset add f2b-sshd 2001:db8:: timeout 600 -exist
ipset v6.29: Syntax error: cannot parse 2001:db8::: resolving to IPv4 address failed
```

This might be an ipset issue, the version I have is default that comes with CentOS 7.4.

**Comment by adbario** [comment]
Using the same fix as in iptables-ipset-proto6.conf seems to work, by using the familyopt to create correct ipset:
https://github.com/fail2ban/fail2ban/blob/0.10/config/action.d/iptables-ipset-proto6.conf

**Comment by sebres** [comment]
Fixed in current 0.10, see e384acca5f085998174346d82102a370f7a47c05.
Thanks!

**Comment by adbario** [comment]
Great thanks! Any plans for releasing 0.10.2 with latest fixes?
